pages/_app.js

```diff
@@ -7,52 +7,79 @@ import { nanoid } from "nanoid";
 export default function App({ Component, pageProps }) {
   const [shoppingItems, setShoppingItems] = useState(initialShoppingItems);
 
+  const shoppingItemsWithCategoryColor = shoppingItems.map((shoppingItem) => {
+    const category = categories.find(
+      (category) => category.name === shoppingItem.category
+    );
+
+    const backgroundColor = category ? category.color : "white";
+
+    return {
+      ...shoppingItem,
+      backgroundColor,
+    };
+  });
+
+  const purchasedItems = shoppingItemsWithCategoryColor.filter(
+    (shoppingItem) => shoppingItem.isPurchased
+  );
+
+  const unpurchasedItems = shoppingItemsWithCategoryColor.filter(
+    (shoppingItem) => shoppingItem.isPurchased === false
+  );
+
   function handleAddItem(newItem) {
-    setShoppingItems([
+    setShoppingItems((prevItems) => [
       {
         id: nanoid(),
+        isPurchased: false,
         ...newItem,
       },
-      ...shoppingItems,
+      ...prevItems,
     ]);
   }
 
   function handleEditItem(id, newItem) {
-    setShoppingItems(
-      shoppingItems.map((shoppingItem) => {
-        if (shoppingItem.id === id) return { ...shoppingItem, ...newItem };
+    setShoppingItems((prevItems) =>
+      prevItems.map((shoppingItem) => {
+        if (shoppingItem.id === id) {
+          return { ...shoppingItem, ...newItem };
+        }
         return shoppingItem;
       })
     );
   }
 
   function handleDeleteItem(id) {
-    setShoppingItems(
-      shoppingItems.filter((shoppingItem) => shoppingItem.id !== id)
+    setShoppingItems((prevItems) =>
+      prevItems.filter((shoppingItem) => shoppingItem.id !== id)
     );
   }
 
-  const shoppingItemsWithCategoryColor = shoppingItems.map((shoppingItem) => {
-    const category = categories.find(
-      (category) => category.name === shoppingItem.category
+  function handleToggleIsPurchased(id) {
+    setShoppingItems((shoppingItems) =>
+      shoppingItems.map((shoppingItem) =>
+        shoppingItem.id === id
+          ? { ...shoppingItem, isPurchased: !shoppingItem.isPurchased }
+          : shoppingItem
+      )
```

Comment:
It's a convention to fix the previous value of the state with "prev":

```suggestion
    setShoppingItems((prevShoppingItems) =>
      prevShoppingItems.map((prevShoppingItem) =>
        prevShoppingItem.id === id
          ? { ...prevShoppingItem, isPurchased: !prevShoppingItem.isPurchased }
          : prevShoppingItem
      )
```

Reply:
Done